src/FastConv/FastConv.cpp

```diff
@@ -1,32 +1,192 @@
 
+#include <cassert>
+#include <iostream>
+
 #include "FastConv.h"
+#include "Fft.h"
+#include "Vector.h"
+
 
-CFastConv::CFastConv( void )
-{
+TimeConvolution::TimeConvolution(const float *impulseResponse, int length)
+: impulseResponse(impulseResponse, &impulseResponse[length]), history(length) {
 }
 
-CFastConv::~CFastConv( void )
-{
-    reset();
+void TimeConvolution::process(float *output, const float *input, int length) {
+    for (int i = 0; i < length; i++) {
+        history.putPostInc(input[i]);
+        float acc = 0;
+        for (int j = 0; j < impulseResponse.size(); j++) {
+            acc += impulseResponse[j] * history.get(-j);
+        }
+        output[i] = acc;
+        history.getPostInc();
+    }
+}
+
+int TimeConvolution::getTailLength() const {
+    // NOTE: The tail is the length of the impulse response minus one,
+    // even if the user has provided fewer than `impulseResponse.size()` input samples.
+    return impulseResponse.size() - 1;
 }
 
-Error_t CFastConv::init(float *pfImpulseResponse, int iLengthOfIr, int iBlockLength /*= 8192*/, ConvCompMode_t eCompMode /*= kFreqDomain*/)
-{
 
-    return Error_t::kNoError;
+FreqConvolution::FreqConvolution(const float *impulseResponse, int length, int blockLength)
+: blockLength(blockLength), tailLength(length - 1 + blockLength), numBlocks((int)ceil((float)length / blockLength)),
+  outputBlock(blockLength*2), saved(blockLength), inputBlocks(numBlocks*blockLength*2), impulseResponseBlocks(numBlocks*blockLength*2) {
+    CFft::createInstance(fft);
+    fft->initInstance(blockLength*2, 1, CFft::kWindowHann, CFft::kNoWindow);
+    for (int i = 0; i < numBlocks; i++) {
+        int thisBlockLength = std::min(blockLength, length - i * blockLength);
+        float *impulseResponseBlock = &impulseResponseBlocks[i * blockLength*2];
+        std::copy_n(&impulseResponse[i * blockLength], thisBlockLength, impulseResponseBlock);
+        fft->doFft(impulseResponseBlock, impulseResponseBlock);
+    }
+    // We scale by the FFT length once, here, to avoid doing in repeatedly in `addMultiplySpectra`.
+    CVectorFloat::mulC_I(impulseResponseBlocks.data(), blockLength * 2, impulseResponseBlocks.size());
+}
+
```

Comment:
Using constructor to initialize known IR's FFT is interesting. It will reduce the main process time since IR calculation is not included.